polar-core/src/filter.rs

```diff
@@ -0,0 +1,491 @@
+            // is exactly this instance, so return a filter that matches it.
+            i @ ExternalInstance(_) => Ok(Filter {
+                root: class.to_string(),
+                relations: HashSet::new(),
+                conditions: vec![singleton(Condition(
+                    Field(Proj(class.to_string(), None)),
+                    Compare::Eq,
+                    Imm(i.clone()),
+                ))],
+            }),
+
+            // oops, we don't know how to handle this!
+            _ => input_error(ands.to_polar()),
+        }
+    }
+
+    fn empty(class: &str) -> Self {
+        use {Datum::Imm, Value::Boolean};
+        Self {
+            root: class.to_string(),
+            relations: Default::default(),
+            conditions: vec![singleton(Condition(
+                Imm(Boolean(true)),
+                Compare::Eq,
+                Imm(Boolean(false)),
+            ))],
+        }
+    }
+
+    fn union(self, mut other: Self) -> Self {
+        other.conditions.extend(self.conditions);
+        other.relations.extend(self.relations);
+        other
+    }
+}
+
+#[derive(Default)]
+struct QueryInfo {
+    types: Types,
+    entities: Map<PathVar, TypeName>,
+    conditions: Set<Condition>,
+    relations: Set<Relation>,
+}
+
+impl QueryInfo {
+    /// try to match a type and a field name with a relation
+    fn get_relation(&mut self, typ: &str, dot: &str) -> Option<Relation> {
+        if let Some(Type::Relation {
+            other_class_tag, ..
+        }) = self.types.get(typ).and_then(|map| map.get(dot))
+        {
+            Some(Relation(
+                typ.to_string(),
+                dot.to_string(),
+                other_class_tag.to_string(),
+            ))
+        } else {
+            None
+        }
+    }
+
+    /// turn a pathvar into a projection
+    fn pathvar2proj(&mut self, pv: PathVar) -> PolarResult<Proj> {
+        let PathVar { mut path, var } = pv;
+        let mut pv = PathVar::from(var); // new var with empty path
+                                         // what type is the base variable?
+        let mut typ = match self.get_type(pv.clone()) {
+            Some(typ) => typ,
+            _ => return invalid_state_error(format!("unknown type for `{}`", pv.var)),
+        };
+
+        // the last part of the path is always allowed not to be a relation.
+        // pop it off for now & deal with it in a minute.
+        let field = path.pop();
+
+        // all the middle parts have to be relations, so if we can't find one
+        // then we fail here.
+        for dot in path {
+            match self.get_relation(&typ, &dot) {
+                None => return unregistered_field_error(&typ, &dot),
+                Some(rel) => {
+                    typ = rel.2.clone();
+                    pv.path.push(rel.1.clone());
+                    self.entities.insert(pv.clone(), typ.clone());
+                    self.relations.insert(rel);
+                }
```

Comment:
Here's another way to fix the name problem that you might like better.

```suggestion
                Some(rel @ Relation(_, ref through, ref to)) => {
                    typ = through.clone();
                    pv.path.push(to.clone());
                    self.entities.insert(pv.clone(), typ.clone());
                    self.relations.insert(rel);
                }
```

Even though the fields don't have names in the struct definition, it's now easy to see what they are when you read this code.

Reply:
Did you try building that?